### PR TITLE
consensus/istanbul: add round to previous hash

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
- This PR implements istanbul compatible change item. Hardfork blocknumber is nil, so it is not activated, yet.
- Main idea
  - When selecting next committee set, a seed is provided to select committee set in random. However, ingredient of the seed is only previous block's hash. For this reason, committee set is very similar between two different rounds.
  - To solve this problem, round value is added to the ingredient of the seed.
- This PR have three commits.
  - b7c369166325014e1c580b9817177ef96b9b1fab: revise fork.Rules to panic instead of returning err when hardForkBlockNumberConfig is not initialized.
  - 22e054069eff979a70a0beedb4ff284bd45ba35f: add round to prevHash (related PR https://github.com/klaytn/klaytn/pull/420)
  - 5b11e69b85268c93f16f3124d8a8188b514d5989: some of the tests are broken due to above changes. it fixes them to work properly.
## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yoomee1313** [comment]
Dear reviewers, revision of #997 and #1001 is done. Please take a look.
This PR will be reviewed first, and then we will review the 1001 PR.
cc. @aidan-kwon @jeongkyun-oh @hyochan-brown @winnie-byun @ehnuje

File: fork/fork.go [hunk 1]
```diff
@@ -33,12 +33,12 @@ var (
...
 	if hardForkBlockNumberConfig == nil {
-		return nil, errors.New("hardForkBlockNumberConfig variable is not initialized")
+		panic("fork.Rules should never be called before hardForkBlockNumberConfig initialization.")
 	}
 	rules := hardForkBlockNumberConfig.Rules(blockNumber)
-	return &rules, nil
+	return &rules
 }
 
 // SetHardForkBlockNumberConfig sets values in HardForkConfig if it is not nil.
```
Comment (winnie-byun) [inline fork/fork.go:44]: There is something wrong with the function. 
The config can be set only once, but it is called with different values : 
```
fork.SetHardForkBlockNumberConfig(&params.ChainConfig{IstanbulCompatibleBlock: istanbul.TestBlockHF})
```
```
fork.SetHardForkBlockNumberConfig(&params.ChainConfig{})
```
We don't know which one will be set.

Comment (yoomee1313) [reply fork/fork.go:44]: @winnie-byun It can be called at test.
```
// CAUTIOUS: Calling this function is can be dangerous, so avoid using it except tests
```

File: consensus/istanbul/validator/default_test.go [hunk 2]
```diff
@@ -210,6 +212,8 @@ func TestStickyProposer(t *testing.T) {
 }
 
 func TestDefaultSet_SubList(t *testing.T) {
+	fork.SetHardForkBlockNumberConfig(&params.ChainConfig{IstanbulCompatibleBlock: istanbul.TestBlockHF})
```
Comment (aidan-kwon) [inline consensus/istanbul/validator/default_test.go:215]: @yoomee1313 Could you change back the HardForkBlockNumberConfig with `defer`? This change may affect other test cases.

Comment (yoomee1313) [reply consensus/istanbul/validator/default_test.go:215]: @aidan-kwon wow.. thanks for good idea. I also worried about the problem that the chainConfig was shared between the tests.

File: consensus/istanbul/backend/engine_test.go [hunk 3]
```diff
@@ -76,9 +76,12 @@ func makeCommittedSeals(hash common.Hash) [][]byte {
 // block by one node. Otherwise, if n is larger than 1, we have to generate
 // other fake events to process Istanbul.
 func newBlockChain(n int, items ...interface{}) (*blockchain.BlockChain, *backend) {
+	// some tests use custom test account
+	var key *ecdsa.PrivateKey
+
 	// generate a genesis block
 	genesis := blockchain.DefaultGenesisBlock()
-	genesis.Config = params.TestChainConfig
+	genesis.Config = &params.ChainConfig{ChainID: big.NewInt(1)}
```
Comment (jeongkyun-oh) [inline consensus/istanbul/backend/engine_test.go:84]: Is this okay with other tests? Why `ChainID` shoud be 1?

File: consensus/istanbul/backend/engine_test.go [hunk 4]
```diff
@@ -98,14 +101,16 @@ func newBlockChain(n int, items ...interface{}) (*blockchain.BlockChain, *backen
...
+			key = v
 		}
 	}
 	nodeKeys = make([]*ecdsa.PrivateKey, n)
 	addrs = make([]common.Address, n)
 
 	var b *backend
 	if len(items) != 0 {
-		b = newTestBackendWithConfig(genesis.Config)
+		b = newTestBackendWithConfig(genesis.Config, key)
```
Comment (jeongkyun-oh) [inline consensus/istanbul/backend/engine_test.go:113]: Can you use 48-52 for keys? Do you need to pass `nodekey` specifically?
```go
// These variables are the global variables of the test blockchain.
var (
	nodeKeys []*ecdsa.PrivateKey
	addrs    []common.Address
)
```

File: consensus/istanbul/backend/backend_test.go [hunk 5]
```diff
@@ -695,41 +680,15 @@ func getGovernance(dbm database.DBManager) *governance.Governance {
...
-		rewardbase:        rewards[0],
-		governance:        gov,
-		nodetype:          common.CONSENSUSNODE,
-		rewardDistributor: reward.NewRewardDistributor(gov),
-	}
-	backend.core = istanbulCore.New(backend, backend.config)
-
-	backend.currentView.Store(&istanbul.View{Sequence: big.NewInt(1), Round: big.NewInt(0)})
-	valSet.SetBlockNum(uint64(1))
+	valSet.SetBlockNum(istanbul.TestBlockBeforeHF.Uint64())
```
Comment (jeongkyun-oh) [inline consensus/istanbul/backend/backend_test.go:691]: Is it okay to remove those code fragments? Looks many lines are removed.
Does this test work for its purpose?

File: consensus/istanbul/config.go [hunk 6]
```diff
@@ -44,3 +48,10 @@ var DefaultConfig = &Config{
 	Epoch:          30000,
 	SubGroupSize:   21,
 }
+
+// Test block numbers used for the tests which uses hard fork block number information
+var (
+	TestBlockBeforeHF = big.NewInt(4)
+	TestBlockHF       = big.NewInt(5)
+	TestBlockAfterHF  = big.NewInt(6)
+)
```
Comment (aidan-kwon) [inline consensus/istanbul/config.go:57]: Do they need to be global variables? I think it can be declared on each test file.

File: consensus/istanbul/backend/backend_test.go [hunk 7]
```diff
@@ -66,382 +62,371 @@ var (
 		{Sequence: 0, Round: 12}:  false,
 		{Sequence: 0, Round: 13}:  false,
 		{Sequence: 0, Round: 14}:  false,
+		{Sequence: 5, Round: 4}:   false,
```
Comment (aidan-kwon) [inline consensus/istanbul/backend/backend_test.go:65]: Just wondering, why did you change these?

File: consensus/istanbul/core/commit_test.go [hunk 8]
```diff
@@ -36,6 +38,9 @@ func TestCore_sendCommit(t *testing.T) {
 
 	mockCtrl.Finish()
 
+	// set hardForkBlockNumberConfig
+	fork.SetHardForkBlockNumberConfig(&params.ChainConfig{IstanbulCompatibleBlock: istanbul.TestBlockHF})
```
Comment (aidan-kwon) [inline consensus/istanbul/core/commit_test.go:42]: It needs to be back to the original HardforkBlockNumber also.

File: consensus/istanbul/validator/default_test.go (hunk 2 above)
Comment (aidan-kwon) [reply consensus/istanbul/validator/default_test.go:215]: Uhm... It may not work well. `SetHardForkBlockNumberConfig` calls `once.Do()` internally. Does it need to be called once?

File: consensus/istanbul/validator/weighted_random_test.go [hunk 9]
```diff
@@ -441,32 +443,92 @@ func TestWeightedCouncil_SetSubGroupSize(t *testing.T) {
...
 func TestWeightedCouncil_SubListWithProposer(t *testing.T) {
-	validators := makeTestValidators(testNonZeroWeights)
-	valSet := makeTestWeightedCouncil(testNonZeroWeights)
+	var (
+		validators = makeTestValidators(testNonZeroWeights)
+		prevHash   = crypto.Keccak256Hash([]byte("This is a test"))
+		valSet     = makeTestWeightedCouncil(testNonZeroWeights)
+
+		expectIndexOfSubsetLenTest                     = []int{1, 2, 7, 3, 11, 6, 9, 4, 0, 8, 12, 10}
+		expectIndexOfRoundTestBeforeIstanbulCompatible = [][]int{
```
Comment (aidan-kwon) [inline consensus/istanbul/validator/weighted_random_test.go:452]: It looks fine, but it would be more intuitive if arrays are sorted.

File: consensus/istanbul/validator/weighted_random_test.go [hunk 10]
```diff
@@ -441,32 +443,92 @@ func TestWeightedCouncil_SetSubGroupSize(t *testing.T) {
...
+		viewBeforeHF := &istanbul.View{Sequence: istanbul.TestBlockBeforeHF, Round: big.NewInt(int64(0))}
+		viewAfterHF := &istanbul.View{Sequence: istanbul.TestBlockAfterHF, Round: big.NewInt(int64(0))}
+		assert.Equal(t, expectSubList, valSet.SubList(prevHash, viewBeforeHF), "test Subset length: %d(before istanbulCompatible)", testSubsetLen)
+		assert.Equal(t, expectSubList, valSet.SubList(prevHash, viewAfterHF), "test subset length: %d(after istanbulCompatible)", testSubsetLen)
+	}
+
+	// Round test: various round test
+	valSet.SetBlockNum(1)
+	valSet.SetSubGroupSize(uint64(len(validators) - 1))
+	for round := 0; round < 15; round++ {
```
Comment (aidan-kwon) [inline consensus/istanbul/validator/weighted_random_test.go:519]: `15` can be changed to `len(expectIndexOfRoundTestBeforeIstanbulCompatible)`. 
It may reduce mistakes when a new test array is added. 
It would be better if this code check whether `len(expectIndexOfRoundTestBeforeIstanbulCompatible) == len(expectIndexOfRoundTestAfterIstanbulCompatible)`

File: fork/fork.go [hunk 11]
```diff
@@ -33,12 +33,12 @@ var (
 
 // Rules returns the hard fork information
 // CAUTIOUS: Use it when chainConfig value is not reachable
-func Rules(blockNumber *big.Int) (*params.Rules, error) {
+func Rules(blockNumber *big.Int) *params.Rules {
 	if hardForkBlockNumberConfig == nil {
-		return nil, errors.New("hardForkBlockNumberConfig variable is not initialized")
+		panic("fork.Rules should never be called before hardForkBlockNumberConfig initialization.")
```
[the remark this answers is not in the document]
Comment (yoomee1313) [reply fork/fork.go]: Reflected. Please take a look at fork/fork.go

File: consensus/istanbul/core/commit_test.go (hunk 8 above)
Comment (yoomee1313) [reply consensus/istanbul/core/commit_test.go]: Reflected in new change https://github.com/klaytn/klaytn/pull/997/commits/7b268b64f681d73f4839a0569baa2484aad823b3

File: consensus/istanbul/backend/backend_test.go (hunk 7 above)
Comment (yoomee1313) [reply consensus/istanbul/backend/backend_test.go:65]: Uhm.. I just want to test both situations, before istanbulHF and after istanbulHF. It is originally from old PR which andy made!

File: consensus/istanbul/config.go (hunk 6 above)
Comment (yoomee1313) [reply consensus/istanbul/config.go:57]: As you mentioned in another comment, the global config is accessed simultaneously by multiple test functions. I was trying to solve the problem with those variables :) It is removed with new change!

File: consensus/istanbul/backend/backend_test.go (hunk 5 above)
Comment (yoomee1313) [reply consensus/istanbul/backend/backend_test.go]: I think it's okay to remove those codes because most lines are same with new function of `consensus/istanbul/backend/backend.go`. Also, i have tested and debugged a lot. Don't worry about this :)

File: consensus/istanbul/backend/engine_test.go (hunk 4 above)
Comment (yoomee1313) [reply consensus/istanbul/backend/engine_test.go:118]: I revised newBlockchain because the `key` below is not configurable(it is needed at `Test_GossipSubPeerTargets`). I couldn't think up another way... do you have any suggestion..?
```
func newTestBackendWithConfig(chainConfig *params.ChainConfig, key *ecdsa.PrivateKey) (b *backend) {
	dbm := database.NewDBManager(&database.DBConfig{DBType: database.MemoryDB})
	key, _ := crypto.GenerateKey()
	if chainConfig.Governance.GovernanceMode == "single" {
		// if governance mode is single, set the node key to the governing node.
		chainConfig.Governance.GoverningNode = crypto.PubkeyToAddress(key.PublicKey)
	}
}
```

File: consensus/istanbul/backend/engine_test.go (hunk 3 above)
Comment (yoomee1313) [reply consensus/istanbul/backend/engine_test.go:89]: `ChainID` is also 1 at `params.TestChainConfig`. It's okay with other tests.

File: consensus/istanbul/validator/default_test.go (hunk 2 above)
Comment (yoomee1313) [reply consensus/istanbul/validator/default_test.go]: Yes, once.Do() needs to be removed.. I reconsidered about using it and decided to remove once.Do() because there's already a check logic whether the global config is set or not. It's been deleted and reset the global config at `defer`, so please take a look. By the way, I think the change is too big.. What do you think?

File: fork/fork.go [hunk 12]
```diff
@@ -19,37 +19,47 @@ package fork
...
+
+// GetHardForkBlockNumberConfig returns hardForkBlockNumberConfig
+// Use only in tests
+func GetHardForkBlockNumberConfig() *params.ChainConfig {
+	return hardForkBlockNumberConfig
+}
+
+// ResetHardForkBlockNumberConfig sets nil
+// Use only in tests
+func ResetHardForkBlockNumberConfig() {
```
Comment (aidan-kwon) [inline fork/fork.go:63]: I think "Clear~" is more proper than "Reset~" for this function name.

File: blockchain/blockchain.go [hunk 13]
```diff
@@ -891,6 +891,9 @@ func (bc *BlockChain) Stop() {
 		_ = triedb.TrieNodeCache().Close()
 	}
 
+	// reset hardForkBlockNumberConfig to nil
+	fork.ResetHardForkBlockNumberConfig()
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:895]: Do we need to clear it? `Stop()` closes resources used in `Blockchain` but doesn't clear configuration.

File: blockchain/chain_makers.go [hunk 14]
```diff
@@ -177,8 +178,15 @@ func GenerateChain(config *params.ChainConfig, parent *types.Block, engine conse
...
 			TrieNodeCacheConfig: statedb.GetEmptyTrieNodeCacheConfig(),
 		}
+
+		prevConfig := fork.GetHardForkBlockNumberConfig()
 		blockchain, _ := NewBlockChain(db, cacheConfig, config, engine, vm.Config{})
-		defer blockchain.Stop()
+		defer func() {
+			blockchain.Stop()
+			if prevConfig != nil {
+				fork.SetHardForkBlockNumberConfig(prevConfig)
```
Comment (aidan-kwon) [inline blockchain/chain_makers.go:187]: Do we need this? If `NewBlockchain` is called later, the config will be set with a new value.

File: blockchain/blockchain.go (hunk 13 above)
Comment (yoomee1313) [reply blockchain/blockchain.go]: I added this line because I thought `Stop` method is symmetric with `NewBlockchain`. In NewBlockchain hardForkBlockNumber is configured, so i reset it in Stop().. btw, this part is moved to new PR #1012.

File: blockchain/chain_makers.go (hunk 14 above)
Comment (yoomee1313) [reply blockchain/chain_makers.go]: I searched how GenerateChain is used. It is used  to generate test blocks and receipts. The blockchain created in here is not used outside. 

Also, in many times, generateChain is called when already `NewBlockchain` is running. It means hardForkBlockNumberConfig is already set. So I wanted to restore original value, so i put these lines.

File: fork/fork.go (hunk 12 above)
Comment (yoomee1313) [reply fork/fork.go:57]: Thanks, aidan.. Changed to Clear in new change https://github.com/klaytn/klaytn/pull/997/commits/cc27ea34d041de3bbef823c5ab78365fa73d3b37!

Comment (ehnuje) [reply fork/fork.go:57]: It is still `Reset` to me.. Could you please check? @yoomee1313

File: consensus/istanbul/validator/weighted.go [hunk 15]
```diff
@@ -383,6 +384,9 @@ func (valSet *weightedCouncil) SubListWithProposer(prevHash common.Hash, propose
 
 	// seed will be used to select a random committee
 	seed, err := ConvertHashToSeed(prevHash)
+	if fork.Rules(view.Sequence).IsIstanbul {
+		seed += view.Round.Int64()
+	}
```
Comment (ehnuje) [inline consensus/istanbul/validator/weighted.go:389]: Not related to this change, however, we're maintaining almost identical code below.. (please check consensus/istanbul/validator/default.go )

File: consensus/istanbul/config.go (hunk 6 above)
Comment (ehnuje) [reply consensus/istanbul/config.go:57]: They still exist for now, could you please check?

**Review by ehnuje — COMMENTED** [review]
Some of your changes are not reflected in github (or just for me)
Could you please check?

File: consensus/istanbul/validator/weighted_random_test.go (hunk 9 above)
Comment (yoomee1313) [reply consensus/istanbul/validator/weighted_random_test.go:456]: I think sorting cannot be done..
In SubsetLen test, the test is organized in a structure where valSet is increased by 1. If subsetLen is 2, expectIndexOfRoundTestBeforeIstanbulCompatible[0][0]-[0][1] is used, and if subsetLen is 8, expectIndexOfRoundTestBeforeIstanbulCompatible[0][0]-[0][7] is used. So I think sorting seems to be difficult.

File: consensus/istanbul/validator/weighted.go (hunk 15 above)
Comment (yoomee1313) [reply consensus/istanbul/validator/weighted.go:389]: > It is still Reset to me.. Could you please check?

@ehnuje oh! I noticed that change is missing too.. i fixed it, sorry for confusion!

>  Not related to this change, however, we're maintaining almost identical code below.. (please check consensus/istanbul/validator/default.go )

Can I ask again what your question means..? Both changes are intended..

File: fork/fork.go (hunk 12 above)
Comment (yoomee1313) [reply fork/fork.go]: It is still Reset to me.. Could you please check?

@ehnuje oh! I noticed that change is missing too.. i fixed it, sorry for confusion!

File: blockchain/chain_makers.go (hunk 14 above)
Comment (yoomee1313) [reply blockchain/chain_makers.go]: I'll reconsider it about this, thanks!

File: blockchain/blockchain.go (hunk 13 above)
Comment (yoomee1313) [reply blockchain/blockchain.go]: I'll reconsider it about this one more time, thanks!

File: consensus/istanbul/validator/weighted_random_test.go (hunk 10 above)
Comment (yoomee1313) [reply consensus/istanbul/validator/weighted_random_test.go]: reflected at https://github.com/klaytn/klaytn/pull/997/commits/d2f24939698fe688dba58022ea14ff1aacb17d5e, thanks!

**Review by aidan-kwon — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]

**Review by winnie-byun — APPROVED** [review]
